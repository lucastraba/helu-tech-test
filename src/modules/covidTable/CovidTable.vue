<template>
  <div class="results-table">
    <FilterBar
      :is-in-graph-view="isInGraphView"
      :suggestion="suggestion"
      @toggle-view="toggleView"
      @filtered="filter = arguments[0]"
    />
    <DataTable
      v-show="!isInGraphView"
      :data="data"
      :filter="filter"
      :loading="loading"
      :error-msg="errorMsg"
      @suggestion="updateSuggestion"
    />
    <Chart
      v-show="isInGraphView"
      :data="data"
      :filter="filter"
      :loading="loading"
      :error-msg="errorMsg"
    />
  </div>
</template>

<script>
import DataTable from './_components/DataTable';
import FilterBar from './_components/FilterBar';
import Chart from './_components/Chart';

export default {
  name: 'MainTable',
  components: {
    DataTable,
    FilterBar,
    Chart
  },
  data () {
    return {
      data: [],
      loading: false,
      errorMsg: '',
      filter: '',
      isInGraphView: false,
      suggestion: ''
    };
  },
  created () {
    this._performSearch();
  },
  methods: {
    toggleView () {
      this.isInGraphView = !this.isInGraphView;
    },
    updateSuggestion (suggestion) {
      this.suggestion = suggestion;
    },
    async _performSearch () {
      this.errorMsg = '';
      this.loading = true;
      const results = await this.$restActions.getAllCountriesData();
      if (results.error) {
        this.errorMsg = results.error;
      } else if (results && results.data && results.data.length) {
        this.data = results.data;
      } else {
        this.errorMsg = 'Something went wrong.';
      }
      this.loading = false;
    }
  }
};
</script>
