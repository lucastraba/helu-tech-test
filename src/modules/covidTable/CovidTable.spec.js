import { createLocalVue, shallowMount } from '@vue/test-utils';
import CovidTable from './CovidTable';

describe('CovidTable Component', () => {
  const mock$restActions = {
    getAllCountriesData: jest.fn().mockReturnValue({
      data: []
    })
  };
  const localVue = createLocalVue();
  const factory = () => shallowMount(CovidTable, {
    localVue,
    stubs: {
      FilterBar: true,
      DataTable: true,
      Chart: true
    },
    mocks: {
      $restActions: mock$restActions
    }
  });
  describe('base rendering', () => {
    it('should render the nav bar in table mode', () => {
      // Arrange.
      const wrapper = factory();
      const table = wrapper.find('datatable-stub');
      const chart = wrapper.find('chart-stub');
      // Act.
      // Assert.
      expect(wrapper.find('filterbar-stub').exists()).toBe(true);
      expect(table.exists()).toBe(true);
      expect(table.attributes('style')).toBeFalsy();
      expect(chart.exists()).toBe(true);
      expect(chart.attributes('style')).toBe('display: none;');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
  describe('conditional rendering', () => {
    it('should render the nav bar in graph mode', async () => {
      const wrapper = factory();
      const table = wrapper.find('datatable-stub');
      const chart = wrapper.find('chart-stub');
      // Act.
      await wrapper.vm.toggleView();
      // Assert.
      expect(table.attributes('style')).toBe('display: none;');
      expect(chart.attributes('style')).toBeFalsy();
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('should render with error message if something failed in the service', async () => {
      // Arrange.
      mock$restActions.getAllCountriesData.mockReturnValueOnce({
        error: 'error'
      });
      const wrapper = factory();
      const table = wrapper.find('datatable-stub');
      const chart = wrapper.find('chart-stub');
      // Act.
      await wrapper.vm.$nextTick();
      // Assert.
      expect(table.attributes('errormsg')).toBe('error');
      expect(chart.attributes('errormsg')).toBe('error');
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('should render with generic error message if no error is provided', async () => {
      // Arrange.
      mock$restActions.getAllCountriesData.mockReturnValueOnce({
        foo: 'bar'
      });
      const wrapper = factory();
      const table = wrapper.find('datatable-stub');
      const chart = wrapper.find('chart-stub');
      // Act.
      await wrapper.vm.$nextTick();
      // Assert.
      expect(table.attributes('errormsg')).toBe('Something went wrong.');
      expect(chart.attributes('errormsg')).toBe('Something went wrong.');
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('should pass the data if everything went 200 OK', async () => {
      // Arrange.
      mock$restActions.getAllCountriesData.mockReturnValueOnce({
        data: ['foo']
      });
      const wrapper = factory();
      const table = wrapper.find('datatable-stub');
      const chart = wrapper.find('chart-stub');
      // Act.
      await wrapper.vm.$nextTick();
      // Assert.
      expect(table.attributes('data')).toBe('foo');
      expect(chart.attributes('data')).toBe('foo');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
  describe('methods', () => {
    describe('updateSuggestion', () => {
      it('should update the suggestion', () => {
        // Arrange.
        const wrapper = factory();
        // Act.
        // Assert.
        expect(wrapper.vm.suggestion).toBe('');
        wrapper.vm.updateSuggestion('foo');
        expect(wrapper.vm.suggestion).toBe('foo');
      });
    });
  });
});
