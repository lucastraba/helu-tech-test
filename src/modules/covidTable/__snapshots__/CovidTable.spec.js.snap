// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CovidTable Component base rendering should render the nav bar in table mode 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"\\" errormsg=\\"\\" loading=\\"true\\"></datatable-stub>
  <chart-stub data=\\"\\" errormsg=\\"\\" loading=\\"true\\" filter=\\"\\" style=\\"display: none;\\"></chart-stub>
</div>"
`;

exports[`CovidTable Component conditional rendering should pass the data if everything went 200 OK 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"foo\\" errormsg=\\"\\"></datatable-stub>
  <chart-stub data=\\"foo\\" errormsg=\\"\\" filter=\\"\\" style=\\"display: none;\\"></chart-stub>
</div>"
`;

exports[`CovidTable Component conditional rendering should render the nav bar in graph mode 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\" isingraphview=\\"true\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"\\" errormsg=\\"Something went wrong.\\" style=\\"display: none;\\"></datatable-stub>
  <chart-stub data=\\"\\" errormsg=\\"Something went wrong.\\" filter=\\"\\" style=\\"\\"></chart-stub>
</div>"
`;

exports[`CovidTable Component conditional rendering should render with error message if something failed in the service 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"\\" errormsg=\\"error\\"></datatable-stub>
  <chart-stub data=\\"\\" errormsg=\\"error\\" filter=\\"\\" style=\\"display: none;\\"></chart-stub>
</div>"
`;

exports[`CovidTable Component conditional rendering should render with generic error message if no error is provided 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"\\" errormsg=\\"Something went wrong.\\"></datatable-stub>
  <chart-stub data=\\"\\" errormsg=\\"Something went wrong.\\" filter=\\"\\" style=\\"display: none;\\"></chart-stub>
</div>"
`;

exports[`NavBar Component base rendering should render the nav bar in table mode 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"\\" errormsg=\\"\\" loading=\\"true\\"></datatable-stub>
  <chart-stub data=\\"\\" errormsg=\\"\\" loading=\\"true\\" suggestion=\\"\\" filter=\\"\\" style=\\"display: none;\\"></chart-stub>
</div>"
`;

exports[`NavBar Component conditional rendering should pass the data if everything went 200 OK 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"foo\\" errormsg=\\"\\"></datatable-stub>
  <chart-stub data=\\"foo\\" errormsg=\\"\\" suggestion=\\"\\" filter=\\"\\" style=\\"display: none;\\"></chart-stub>
</div>"
`;

exports[`NavBar Component conditional rendering should pass the data if everything went 200 OK 2`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"foo\\" errormsg=\\"\\"></datatable-stub>
  <chart-stub data=\\"foo\\" errormsg=\\"\\" suggestion=\\"\\" filter=\\"\\" style=\\"display: none;\\"></chart-stub>
</div>"
`;

exports[`NavBar Component conditional rendering should render the nav bar in graph mode 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\" isingraphview=\\"true\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"\\" errormsg=\\"Something went wrong.\\" style=\\"display: none;\\"></datatable-stub>
  <chart-stub data=\\"\\" errormsg=\\"Something went wrong.\\" suggestion=\\"\\" filter=\\"\\" style=\\"\\"></chart-stub>
</div>"
`;

exports[`NavBar Component conditional rendering should render with error message if something failed in the service 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"\\" errormsg=\\"error\\"></datatable-stub>
  <chart-stub data=\\"\\" errormsg=\\"error\\" suggestion=\\"\\" filter=\\"\\" style=\\"display: none;\\"></chart-stub>
</div>"
`;

exports[`NavBar Component conditional rendering should render with generic error message if no error is provided 1`] = `
"<div class=\\"results-table\\">
  <filterbar-stub suggestion=\\"\\"></filterbar-stub>
  <datatable-stub filter=\\"\\" data=\\"\\" errormsg=\\"Something went wrong.\\"></datatable-stub>
  <chart-stub data=\\"\\" errormsg=\\"Something went wrong.\\" suggestion=\\"\\" filter=\\"\\" style=\\"display: none;\\"></chart-stub>
</div>"
`;
